app/grants/views.py

```diff
@@ -0,0 +1,430 @@
+
+    if request.method == "POST":
+        method = request.POST.get('method')
+
+        if method == "POST":
+            form = MilestoneForm(request.POST)
+            milestone = form.save(commit=False)
+            milestone.grant = grant
+            milestone.save()
+
+        if method == "PUT":
+            milestone_id = request.POST.get('milestone_id')
+            milestone = Milestone.objects.get(pk=milestone_id)
+            milestone.completion_date = request.POST.get('completion_date')
+            milestone.save()
+
+        if method == "DELETE":
+            milestone_id = request.POST.get('milestone_id')
+            milestone = grant.milestones.get(pk=milestone_id)
+            milestone.delete()
+
+        return redirect(reverse('grants:milestones', args=(grant.pk, grant.slug)))
+
+    form = MilestoneForm()
+    milestones = grant.milestones.order_by('due_date')
+
+    params = {
+        'active': 'grant_milestones',
+        'title': _('Grant Milestones'),
+        'card_desc': _('Provide sustainable funding for Open Source with Gitcoin Grants'),
+        'grant': grant,
+        'milestones': milestones,
+        'form': form,
+        'keywords': get_keywords(),
+    }
+    return TemplateResponse(request, 'grants/milestones.html', params)
+
+
+@login_required
+def grant_fund(request, grant_id,  grant_slug):
+    """Handle grant funding."""
+    try:
+        grant = Grant.objects.get(pk=grant_id, slug=grant_slug)
+    except Grant.DoesNotExist:
+        raise Http404
+
+    profile = request.user.profile if request.user.is_authenticated and request.user.profile else None
+
+    if not grant.active:
+        params = {
+            'active': 'grant_error',
+            'title': _('Grant Ended'),
+            'grant': grant,
+            'text': _('This Grant is not longer active.')
+        }
+        return TemplateResponse(request, 'grants/shared/error.html', params)
+
+    if grant.admin_profile == profile:
+        params = {
+            'active': 'grant_error',
+            'title': _('Invalid Grant Subscription'),
+            'grant': grant,
+            'text': _('You cannot fund your own Grant.')
+        }
+        return TemplateResponse(request, 'grants/shared/error.html', params)
+
+    active_subscription = Subscription.objects.select_related('grant').filter(
+        grant=grant_id, active=True, contributor_profile=request.user.profile
+    )
+
+    if active_subscription:
+        params = {
+            'active': 'grant_error',
+            'title': _('Subscription Exists'),
+            'grant': grant,
+            'text': _('You already have an active subscription for this grant.')
+        }
+        return TemplateResponse(request, 'grants/shared/error.html', params)
+
+    # make sure a user can only create one subscription per grant
+    if request.method == 'POST':
+        subscription = Subscription()
+
+        subscription.subscription_hash = request.POST.get('subscription_hash', '')
```

Comment:
is it worth doing some validation on the POST?

Reply:
Yes, probably.